### PR TITLE
Don 154/donation webhook usage refactor

### PR DESCRIPTION
**Pull request by jgreever**
# Setup instructions at the bottom

Nuke all your old docker images for WonderTix before using this one. Since I don't have anything else on my docker except WonderTix, I use: `docker system prune --all --force; docker builder prune --all --force` as this clears docker to a fresh state and removes any old build information so we are getting all new images created.

Make sure to edit the main `.env` file and add `MINIMAL_SEED=true` at the bottom so the entire DB doesn't need to be seeded, only minimally needed data (This is from Jaeger and Dylan and the work they did to minimize the seeding being merged into this branch already.)

### Description
The donation endpoint has been finished and refactored. It now accepts the donation, creates or updates a contact in the system, then creates the donation and adds it to the DB.

Currently, there is no method in place to refund a donation, but it can be added. The front end part has commented out the code for recurring donations, as that will require working with PPH admins and accessing the Stripe dashboard (PII in there) and creating the recurring donations amounts. This will be documented and handed to the next team to work with PPH admins on.

This branch has all code currently from main merged into it. There may be quite a bit of linting and prettier changes that happened on some files, but please take care and actually download this branch and run it locally against your own stripe account to validate the workflow is correct.

### Risks / Issues
Not all testing has been done on a stripe donation. There  could be a crash if the value is too high (happened once to me, but wasn't able to reproduce it to verify it).

Refunds will need to be added at some point so they do not have to be done via the Stripe dashboard.

Because of the nature of merging, there may be changes in here that are from other branches that may not be fully merged in yet. I did my best to convert those MRs to a draft and inform the developer that I have already merged their code into my branch. That I am aware of, there is only one MR: https://github.com/WonderTix/WonderTix/pull/413 via branch: https://github.com/WonderTix/WonderTix/tree/DON-188 and they have been notified and the MR commented and converted to draft already. This was done because modifications at the last minute needed to be both done and tested and unfortunately we were not able to follow standard branching and merging steps.

### Validation
Verified locally and via Stripe dashboard and in the DB (contacts/donations)

No playwright tests exist for the donation process yet, but can be written if needed.

### Operating System
Windows 11 (AMD, x86_64)
macOS 15 (Intel, x86_64)

### Setup
First, follow the `README.md` file's section for basic setup ([right here](https://github.com/WonderTix/WonderTix#setup)). If you haven't setup a Stripe account, you will need to do that or obtain a testing key from a team lead.

Second, bring up the system with the typical `docker-compose up -d` and wait for everything to spin up.

Once the client/server/db are up and ready, visit the site: [https://localhost:3000/](https://localhost:3000) and click `Donate` in the top navigation bar.

Either select a preset amount, or enter the amount you want to donate (not to exceed $99,999, as that is Stripes limit and it can cause a crash at times), fill out the form and click `Next`.

You will now be at the stripe checkout page. You can use the following test card data:

CC#: `4242 4242 4242 4242`
Exp Date: `10\29`
CCV: `789`

You can fill out the rest of the information with whatever you want, it's a test transaction so it doesn't matter what you put in for email/zip/etc, just the payment information needs to be valid and the above comes from Stripes documentation found [here](https://stripe.com/docs/testing).

After submitting, you should see a success page, then a button to go home.

Now if you check the database using either PostgreSQL client, or go in the server directory and type `npx prisma studio`, you can view the donation and contacts table to see the new donation and contact. It will either update a contact with the new information, or create a new contact. It goes by the email address. Donations are always a new entry, but get tied to the contact.

Please reach out to me: jgreever@pdx.edu or via Slack if you have any questions while setting this up.

### CODE REVIEW

File: server/src/controllers/orderController.ts
Comment (benfaustin95) [inline server/src/controllers/orderController.ts:49]: It might be worth checking the checkout session status before recording the transaction, just in case it failed.

File: server/src/controllers/orderController.ts
Comment (benfaustin95) [inline server/src/controllers/orderController.ts:34]: It looks like the donation is not being handled when it is made during checkout, should that be added to the webhook or should that option be entirely removed from the checkout process?

File: server/src/controllers/orderController.ts
Comment (jgreever) [reply server/src/controllers/orderController.ts:49]: I will get that added in this weekend and thank you for reviewing this MR for me.

File: server/src/controllers/orderController.ts
Comment (jgreever) [reply server/src/controllers/orderController.ts:34]: Donation process should be separate from ticket purchasing. I understand that it would be nice to buy a ticket AND donate at the same time, but we could never get that working correctly, unfortunately. If we can split them, I'm more than happy to keep looking into making it possible to buy tickets and donate in the same transaction.

Comment (benfaustin95) [reply server/src/controllers/orderController.ts:34]: Currently the checkout process has an option to tack on a donation to a ticket purchase so that might need to be removed.

File: server/src/controllers/orderController.ts
Comment (benfaustin95) [reply server/src/controllers/orderController.ts:49]: Another thing to take a look at would be leaving the refund intent null unless the donation has been refunded.

File: server/src/controllers/orderController.ts
Comment (ben16williams) [reply server/src/controllers/orderController.ts:34]: We will want want to have the ability to both make a donation while purchasing a ticket (in the same transaction) and make a separate donation. If we have to split the webhook or make a new one to support both cases then that should be fine

Comment (benfaustin95) [reply server/src/controllers/orderController.ts:34]: I can take a look at incorporating the donation record creation with the ticketing webhook session type.

Comment (ben16williams) [reply server/src/controllers/orderController.ts:34]: That would be great! If you do, would you mind creating a ticket on the project board?

**Comment by jgreever** [comment]
Take a look at the changes I have made to the donation webhook side of things and let me know if you find any issues that need fixed.

File: client/src/components/Ticketing/ticketingmanager/Refund/RefundOrderTicket.tsx
Comment (benfaustin95) [inline client/src/components/Ticketing/ticketingmanager/Refund/RefundOrderTicket.tsx:1]: I think this is an old file that no longer exists

File: client/src/components/Ticketing/donation/DonorBenefitsPage.tsx
Comment (benfaustin95) [inline client/src/components/Ticketing/donation/DonorBenefitsPage.tsx:39]: Would it be possible to go through and update divs that only contain text to p tags so that they are a little more semantic?

File: client/src/components/Ticketing/donation/OnlyDonatePage.tsx
Comment (benfaustin95) [inline client/src/components/Ticketing/donation/OnlyDonatePage.tsx:111]: Would it be possible to make this a void tag since the body is empty?

File: client/src/components/Ticketing/donation/OnlyDonatePage.tsx
Comment (benfaustin95) [inline client/src/components/Ticketing/donation/OnlyDonatePage.tsx:47]: Might be worth adding a pop up to communicate if checkout failed to the user, or maybe adding it as an issue to implement in the future.

File: client/src/components/Ticketing/donation/OnlyDonatePage.tsx
Comment (jgreever) [reply client/src/components/Ticketing/donation/OnlyDonatePage.tsx:111]: I'm not the front end guy who designed this so, no. They are no longer in the capstone. You need to assign this to a new student in the capstone to work on as I only do the backend part, and I am working as a volunteer at this point as I finished this last term.

Please assign this front end tasks to a new capstone member please and thank you.

Comment (ben16williams) [reply client/src/components/Ticketing/donation/OnlyDonatePage.tsx:111]: Thanks for following up on this Justin. I'll address Ben's comment so that we can close out this PR.

File: client/src/components/Ticketing/donation/DonorBenefitsPage.tsx
Comment (ben16williams) [reply client/src/components/Ticketing/donation/DonorBenefitsPage.tsx:39]: This page actually looks like it's trying to reconstruct this [existing page](https://portlandplayhouse.org/donor-benefits/) on Portland Playhouse's website but I can't get it to load in wondertix. PPH can just keep the page that they have on their actual site so we don't need this

File: client/src/components/Ticketing/donation/OnlyDonatePage.tsx
Comment (ben16williams) [reply client/src/components/Ticketing/donation/OnlyDonatePage.tsx]: Made that a void tag https://github.com/WonderTix/WonderTix/pull/517/commits/48dbdd942b1b06e07e18ea566bf397b636f546cd

File: client/src/components/Ticketing/donation/DonorBenefitsPage.tsx
Comment (ben16williams) [reply client/src/components/Ticketing/donation/DonorBenefitsPage.tsx]: Removed https://github.com/WonderTix/WonderTix/pull/517/commits/f9635fba6475e863eaf08301b78c89025bd99195

File: client/src/components/Ticketing/donation/OnlyDonatePage.tsx
Comment (ben16williams) [reply client/src/components/Ticketing/donation/OnlyDonatePage.tsx]: I think this is what you meant? Fair warning I'm not that familiar with front-end elements, including html https://github.com/WonderTix/WonderTix/pull/517/commits/48dbdd942b1b06e07e18ea566bf397b636f546cd

File: client/src/components/Ticketing/donation/OnlyDonatePage.tsx
Comment (ben16williams) [reply client/src/components/Ticketing/donation/OnlyDonatePage.tsx:46]: Created a ticket https://github.com/WonderTix/WonderTix/issues/539

**Comment by ben16williams** [comment]
I've addressed @benfaustin95 and will wait for their approval 🙏  before otherwise merging this down in a day or so

**Review by benfaustin95 — APPROVED** [review]

**Review by ben16williams — APPROVED** [review]

**Pull request merged**
